### ISSUE TITLE
Bindgen incorrectly generates pointers to block pointers.

### ISSUE THREAD

**Issue by gmnicke2** [rust-lang/rust-bindgen#1580]
### Input C/C++ Header

```C
struct contains_ptr_to_block_ptr {
    void (^*val)(int);
};
```

### Bindgen Invocation

```
$ bindgen input.h --generate-block -- -fblocks
```

### Actual Results

```rust
use block;
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct contains_ptr_to_block_ptr {
    pub val: _bindgen_ty_id_4,
}
#[test]
fn bindgen_test_layout_contains_ptr_to_block_ptr() {
    assert_eq!(
        ::std::mem::size_of::<contains_ptr_to_block_ptr>(),
        8usize,
        concat!("Size of: ", stringify!(contains_ptr_to_block_ptr))
    );
    assert_eq!(
        ::std::mem::align_of::<contains_ptr_to_block_ptr>(),
        8usize,
        concat!("Alignment of ", stringify!(contains_ptr_to_block_ptr))
    );
    assert_eq!(
        unsafe {
            &(*(::std::ptr::null::<contains_ptr_to_block_ptr>())).val
                as *const _ as usize
        },
        0usize,
        concat!(
            "Offset of field: ",
            stringify!(contains_ptr_to_block_ptr),
            "::",
            stringify!(val)
        )
    );
}
pub type _bindgen_ty_id_4 = *const ::block::Block<(::std::os::raw::c_int,), ()>;
```

### Expected Results

Bindgen should produce a struct whose field is a mutable pointer to the block pointer:

```rust
pub struct contains_ptr_to_block_ptr {
    pub val: *mut _bindgen_ty_id_4,
}
```

**Comment by emilio** [comment]
So seems like all block pointers are const at the moment:

https://github.com/rust-lang/rust-bindgen/blob/8779c9167f1bb8429f8a13b046be52d5c2654e06/src/codegen/mod.rs#L4007

We should be tracking constness of them already, so this should be trivial to fix (see the usages of `Type::is_const` over there).

Wanna give this a shot? :)

**Comment by gmnicke2** [comment]
Thanks for the response @emilio .

If I'm understanding correctly, I think the permanent constness of the block pointer itself is fine, as it should exist as an immutable pointer to a block. What's as issue here, I believe, is that both a block pointer as a field, and a pointer to a block pointer as a field, both end up as `*const ::block::Block`, while the latter should be `*mut *const ::block::Block` after following aliases.

If I have these two C structs in the header
```C
struct x {
    void (^val)(int);
};

struct y {
    void (^*ptr_val)(int);
};
```

Bindgen is producing the following (after following the `_bindgen_ty_id_xxx` aliases), noting the fields are the same:
```rust
pub struct x {
    pub val: *const ::block::Block<(::std::os::raw::c_int,), ()>;
}

pub struct y {
    pub ptr_val: *const ::block::Block<(::std::os::raw::c_int,), ()>;
}
```

I've looked into it a bit, and it seems like maybe we want to treat `BlockPointer` the way `Pointer` is treated at the IR level, and have its canonical type just be itself (`Some(self)`):
https://github.com/rust-lang/rust-bindgen/blob/8779c9167f1bb8429f8a13b046be52d5c2654e06/src/ir/ty.rs#L348-L358

Apologies if the initial issue statement was unclear or if I misunderstood your guidance. I have a local patch that fixes and tests this this way, if the above logic seems sane enough.

@highfive: assign me

**Comment by emilio** [comment]
Oh you're right, sorry. I misread the example.

But you're correct that that canonical type bit is wrong. It's treating blocks the same way as aliases, while it should do that the same way as pointers.

That change sounds good, so if you submit a PR with that and a test I'm happy to merge it.

Thank you!

**Comment by emilio** [comment]
Fixed by #1582, thanks!
